### ISSUE TITLE
Opening compressed files and archives in JupyterLab

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#8516]
### Problem

* I often work with large files. I think that many users in the target audience of JupyterLab work with such files. These files often come compressed as .gz files.
* I often find myself in a need to have a quick look at the file. My typical workflow is to `zcat | head`, `zcat | tail` and if I need anything more `vim`. Vim handles both archives (zip) and (most of) compressed files (.gz) well.
* I would like to be able to use the JupyterLab GUI to have a quick peek into such files without having to go to the terminal because it would save some time and give me a better insight into tabular files.
* Ideally this would support password-protected zip files too. Those are difficult to access at the moment.

### Proposed Solution

* The `FileEditor` and the `CSVViewer` should open gzip files. Currently, we see an error: `file.gz is not UTF-8 encoded`
   * could this be implemented in the content manager?
   * on the backend this could be implemented with [Python gzip](https://docs.python.org/3/library/gzip.html) module.
* The file browser should open the zip and tar archives.
  * it may be beneficial to indicate that the user is in an archive
  * it could be done in a separate extension which would target archives
  * I would argue that this should be a built-in to keep this function secure and up-to-date
  * on the backend this can be implemented with [Python ZipFile](https://docs.python.org/3/library/zipfile.html) which means we get good cross-platform compatibility for free.
* I do not propose to enable full editing capabilities on compressed files or archives. Just having a tool to view the files would be great. Editing would be perfect but probably is less needed (and may require more work).

### Additional context

* There is [jupyter-archive](https://github.com/hadim/jupyter-archive), an interesting extension which allows compressing and downloading files from the file browser; it allows uncompressing files from the file browser too (see #7477). This is not what I want because uncompressing my files is not feasible:
  - I would run out of disk space; many genomic files use staggeringly repetitive formats which result in extreme compression rates (it is easy to get hundreds of GB into a gzip with less than few hundreds of MB).
  - some zips have password-protected files
* Google services (Drive/Gmail) are an example of services which open zip archives for the user rather than just download the zip.

**Comment by krassowski** [comment]
@JDLEarley I read that you are working on the filebrowser experience, and wanted to highlight this issue for your considertaion.

**Comment by jasongrout** [comment]
Keep in mind that (right now) the entire file has to be transferred to the browser for you to be able to see it, so if it really is a large file (e.g., larger than memory), it will cause issues.

**Comment by JDLEarley** [comment]
Thanks @jasongrout and @krassowski! A potential "preview file without opening it" feature request has come up repeatedly in our research so far, and this is another solid data point on that front.

**Comment by jasongrout** [comment]
I would love to see smarter logic on the python side to be able to support summaries or previews of files without having to transfer the whole file over, or support file ranges to be able to stream file content on demand. We've had many experiments over the years with this sort of thing. I think the most promising were experiments by @danielballan with his hdf5 reader, which allowed server plugins to create new types of responses from the server for specific files.

**Comment by jasongrout** [comment]
> Keep in mind that (right now) the entire file has to be transferred to the browser for you to be able to see it, so if it really is a large file (e.g., larger than memory), it will cause issues.

However, I should add that I don't think this should stop people from exploring this. Being able to open 10MB zip files would still be really useful, I think.

I think .gz files may be the easiest to support. For example, the file contents manager could transparently present .gz files as uncompressed files, and JupyterLab would never know the difference (though it may be confusing to users to see `myfile.csv.gz` as `myfile.csv` in the ux, and you'd have to deal with name conflicts since there could be another `myfile.csv` file actually there). The file contents manager could also transparently present .zip files as directories. Both of these scenarios would involve implementing a server plugin to override the file contents manager, and have nothing to do with JupyterLab per se, but are Jupyter Notebook server plugins.

**Comment by krassowski** [comment]
Thank you for the valuable feedback. I would say that the JupyterLab may need small tweaks if we would want to allow the opening of password-protected files (a simple password pop-up).

I was originally thinking about hdf5 style extension for the zip archive functionality which would create a separate sidebar tab, but maybe having this as a server extension which is mostly transparent would be better.

Two quick questions:
- can we have multiple content managers at once?
- I thought that [jupyterlab-hdf5](https://github.com/jupyterlab/jupyterlab-hdf5) can stream on-demand. is this the case or just my misperception?

**Comment by jasongrout** [comment]
> can we have multiple content managers at once?

No, but what Dan implemented was a way to extend the existing file contents manager



>  * I thought that [jupyterlab-hdf5](https://github.com/jupyterlab/jupyterlab-hdf5) can stream on-demand. is this the case or just my misperception?

Yes, but it is implemented by setting up another connection to the server outside of the notebook file API.

**Comment by krassowski** [comment]
Another nice (small) tweak in lab would be having an icon of an archive for the zip archives - if we implement this as a content manager then the directories and archives would both be presented the same (unless we add a flag in response which tells the frontend that something is an archive).
